cannon/mipsevm/memory/hasher.go

```diff
@@ -0,0 +1,74 @@
+package memory
+
+import (
+	"sync"
+
+	"github.com/ethereum/go-ethereum/crypto"
+)
+
+// Byte32Pool is a sync.Pool for [32]byte slices
+var Byte32Pool = sync.Pool{
+	New: func() interface{} {
+		var b [32]byte
+		return &b // Return a pointer to avoid extra allocations
+	},
+}
+
+// GetByte32 retrieves a *[32]byte from the pool
+func GetByte32() *[32]byte {
+	return Byte32Pool.Get().(*[32]byte)
+}
+
+// PutByte32 returns a *[32]byte to the pool
+func PutByte32(b *[32]byte) {
+	// Optional: Zero the array before putting it back
+	*b = [32]byte{}
+	Byte32Pool.Put(b)
+}
+
+var hashPool = sync.Pool{
+	New: func() interface{} {
+		return crypto.NewKeccakState()
+	},
+}
+
+func GetHasher() crypto.KeccakState {
+	return hashPool.Get().(crypto.KeccakState)
+}
+
+func PutHasher(h crypto.KeccakState) {
+	h.Reset()
+	hashPool.Put(h)
+}
+
+func HashPairNodes(out *[32]byte, left, right *[32]byte) {
+	h := GetHasher()
+	h.Write(left[:])
+	h.Write(right[:])
+	h.Read(out[:])
```

Comment:
The linter doesn't like this ignoring the return value. Given the return value is ignored by the `Keccak256Hash` we were previously using I think we can just silence the linter with:
```suggestion
	_, _ = h.Read(out[:])
```